src/Dependencies/Collections/Internal/ThrowHelper.cs

```diff
@@ -0,0 +1,343 @@
+            throw new InvalidOperationException(GetResourceString(resource), e);
+        }
+
+        [DoesNotReturn]
+        internal static void ThrowNotSupportedException(ExceptionResource resource)
+        {
+            throw new NotSupportedException(GetResourceString(resource));
+        }
+
+        [DoesNotReturn]
+        internal static void ThrowArgumentException_Argument_InvalidArrayType()
+        {
+            throw new ArgumentException(SR.Argument_InvalidArrayType);
+        }
+
+        [DoesNotReturn]
+        internal static void ThrowInvalidOperationException_InvalidOperation_EnumFailedVersion()
+        {
+            throw new InvalidOperationException(SR.InvalidOperation_EnumFailedVersion);
+        }
+
+        [DoesNotReturn]
+        internal static void ThrowInvalidOperationException_InvalidOperation_EnumOpCantHappen()
+        {
+            throw new InvalidOperationException(SR.InvalidOperation_EnumOpCantHappen);
+        }
+
+        [DoesNotReturn]
+        internal static void ThrowInvalidOperationException_ConcurrentOperationsNotSupported()
+        {
+            throw new InvalidOperationException(SR.InvalidOperation_ConcurrentOperationsNotSupported);
+        }
+
+        private static ArgumentException GetArgumentException(ExceptionResource resource)
+        {
+            return new ArgumentException(GetResourceString(resource));
+        }
+
+        private static ArgumentException GetWrongKeyTypeArgumentException(object? key, Type targetType)
+        {
+            return new ArgumentException(string.Format(SR.Arg_WrongType, key, targetType), nameof(key));
+        }
+
+        private static ArgumentException GetWrongValueTypeArgumentException(object? value, Type targetType)
+        {
+            return new ArgumentException(string.Format(SR.Arg_WrongType, value, targetType), nameof(value));
+        }
+
+        private static KeyNotFoundException GetKeyNotFoundException(object? key)
+        {
+            return new KeyNotFoundException(string.Format(SR.Arg_KeyNotFoundWithKey, key));
+        }
+
+        private static ArgumentOutOfRangeException GetArgumentOutOfRangeException(ExceptionArgument argument, ExceptionResource resource)
+        {
+            return new ArgumentOutOfRangeException(GetArgumentName(argument), GetResourceString(resource));
+        }
+
+        // Allow nulls for reference types and Nullable<U>, but not for value types.
+        // Aggressively inline so the jit evaluates the if in place and either drops the call altogether
+        // Or just leaves null test and call to the Non-returning ThrowHelper.ThrowArgumentNullException
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void IfNullAndNullsAreIllegalThenThrow<T>(object? value, ExceptionArgument argName)
+        {
+            // Note that default(T) is not equal to null for value types except when T is Nullable<U>.
+            if (!(default(T) == null) && value == null)
```

Comment:
```suggestion
            if (default(T) != null && value == null)
```

Also, would it not be better to put the `value == null` first? I somewhat assume the JIT can remove it either way, so it would be better to have the relevant test first?

Reply:
This is verbatim from [dotnet/runtime](https://github.com/dotnet/runtime/blob/f56ef95669c115510c0b6dc294634d932fd4e49b/src/libraries/System.Private.CoreLib/src/System/ThrowHelper.cs#L492)